File: lib/sirvlog_client.js

var net = require('net');
var protocol = require('./sirvlog_protocol');
var util = require("util");
var events = require("events");
var os = require("os");
var _ = require('underscore');

exports.Sirvlog = Sirvlog = function (options) {

    this.messageNum = 0;

    this.options = options;

    if (this.options.backlogLimit === undefined) this.options.backlogLimit = 512;
    this.facility = (typeof options.facility === 'string' && options.facility) ? options.facility : 'sirvlog';

    this.hostname = os.hostname();

    this.connected = false;
    this.backlog = {};
    this.connectionAttempt = 1;
    this.init = true;

    //this.messageID = 0;

    this.connect();
};

util.inherits(Sirvlog, events.EventEmitter);

exports.createClient = function (options) {
    return new Sirvlog(options);
};


Sirvlog.prototype.connect = function () {

    var self = this;

    self.socket = net.createConnection(self.options.server.port, self.options.server.address, function () {

            self.protocol = protocol.createProtocol(self.socket).on('error', function (e) {
                self.emit('error', e);
            });

            self.connected = true;

            //console.log('connected to server');

            if (_.size(self.backlog)) { // send all queued logs
                //console.log('send all queued logs, num=', _.size(self.backlog));

                _.each(self.backlog, function (msg, id) {
                    self.protocol.send(JSON.stringify(msg), function (messageId) {
                        //console.log('message ', messageId, ' was sent');
                        delete self.backlog[messageId];
                    }.bind(this, id));
                });
            }

            if (self.init) {
                self.emit('ready'); // to listen in main app
                self.init = false;
            }

            self.connectionAttempt = 1;

        }).on('error', function (e) {

            if (e.code == 'ECONNREFUSED') {
                e.message = 'Can\'t connect to Sirvlog server at ' + self.options.server.address + ':' + self.options.server.port;
            }
            self.emit('error', e);

        }).on('close', function (had_error) {

            self.connected = false;
            if (self.connectionAttempt > 100) self.connectionAttempt = 100;
            setTimeout(self.connect.bind(this), self.init?3000:50 * self.connectionAttempt++);

        });
}


Sirvlog.LEVEL_EMERGENCY = 0;
Sirvlog.LEVEL_ALERT = 1;
Sirvlog.LEVEL_CRITICAL = 2;
Sirvlog.LEVEL_ERROR = 3;
Sirvlog.LEVEL_WARNING = 4;
Sirvlog.LEVEL_NOTICE = 5;
Sirvlog.LEVEL_INFO = 6;
Sirvlog.LEVEL_DEBUG = 7;

Sirvlog.prototype._store = function(logObj){

    var self = this;

    if(!logObj.hostname){
        logObj.hostname = self.hostname;
    }

    if(!logObj.level){
        logObj.level = Sirvlog.LEVEL_INFO;
    }

    if(!logObj.facility){
        logObj.facility = self.facility;
    }

    logObj.order = self.messageNum++;

    if (self.options.backlogLimit) {
        if (_.size(self.backlog) > self.options.backlogLimit) {
            var msgId = _.keys(self.backlog)[0];
            var logObj1 = self.backlog[msgId];
            console[logObj1.level > 3 ? 'log' : 'error'](new Date(), logObj1); // log this message and drop it from buffer
            delete self.backlog[msgId];
        }
        self.backlog[logObj.order] = logObj; // put message to backlog
    }

    if (!self.connected) {
        if (!self.options.backlogLimit) {
            console[logObj.level > 3 ? 'log' : 'error'](new Date(), logObj); // log this message
        }
        return;
    }

    //console.log(self.backlog);

    self.protocol.send(JSON.stringify(logObj), function () {
        if(self.backlog[logObj.order]){
            delete self.backlog[logObj.order];
        }
    });
}

Sirvlog.prototype._log = function (level, message, backtrace, file, line, custom) {

    if (level < Sirvlog.LEVEL_EMERGENCY || level > Sirvlog.LEVEL_DEBUG) level = Sirvlog.LEVEL_INFO;

    if (typeof message !== 'string' || message == '') return;

    var logObj = {
        hostname: this.hostname,
        //timestamp: Date.now(),
        facility: this.facility,
        level: level,
        message: message
    }

    if (typeof custom === 'object') {
        logObj.custom = custom;
    }

    if (typeof line === 'number') {
        logObj.line = line;
    } else if (typeof line === 'object') {
        logObj.custom = line;
    }

    if (typeof file === 'string' && file != '') {
        logObj.file = file;
    } else if (typeof file === 'object') {
        logObj.custom = file;
    }

    if (typeof backtrace === 'string' && backtrace != '') {
        logObj.backtrace = backtrace;
    } else if (typeof backtrace === 'object') {
        logObj.custom = backtrace;
    }

    this._store(logObj);

}

Sirvlog.prototype.log = function (message, backtrace, file, line, custom) {
    this._log(Sirvlog.LEVEL_INFO, message, backtrace, file, line, custom);
}

Sirvlog.prototype.emergency = function (message, backtrace, file, line, custom) {
    this._log(Sirvlog.LEVEL_EMERGENCY, message, backtrace, file, line, custom);
}

Sirvlog.prototype.alert = function (message, backtrace, file, line, custom) {
    this._log(Sirvlog.LEVEL_ALERT, message, backtrace, file, line, custom);
}

Sirvlog.prototype.critical = function (message, backtrace, file, line, custom) {
    this._log(Sirvlog.LEVEL_CRITICAL, message, backtrace, file, line, custom);
}

Sirvlog.prototype.error = function (message, backtrace, file, line, custom) {
    this._log(Sirvlog.LEVEL_ERROR, message, backtrace, file, line, custom);
}

Sirvlog.prototype.warning = function (message, backtrace, file, line, custom) {
    this._log(Sirvlog.LEVEL_WARNING, message, backtrace, file, line, custom);
}

Sirvlog.prototype.notice = function (message, backtrace, file, line, custom) {
    this._log(Sirvlog.LEVEL_NOTICE, message, backtrace, file, line, custom);
}

Sirvlog.prototype.info = function (message, backtrace, file, line, custom) {
    this._log(Sirvlog.LEVEL_INFO, message, backtrace, file, line, custom);
}

Sirvlog.prototype.debug = function (message, backtrace, file, line, custom) {
    this._log(Sirvlog.LEVEL_DEBUG, message, backtrace, file, line, custom);
}

File: ruby/lib/sirvlog_client.rb
# encoding: utf-8

require 'json'
require 'socket'
require 'zlib'
require 'json'
require 'logger'

class SirvlogClient

  LEVEL_EMERGENCY = 0
  LEVEL_ALERT = 1
  LEVEL_CRITICAL = 2
  LEVEL_ERROR = 3
  LEVEL_WARNING = 4
  LEVEL_NOTICE = 5
  LEVEL_INFO = 6
  LEVEL_DEBUG = 7

  ENCODING_PLAIN = 0x01
  ENCODING_LZ77 = 0x02
  ENCODING_DEFLATE = 0x03


  def initialize(opts)

    @backlog = []

    @defaults = {
        'facility' => 'sirvlog', # default logging facility
        'server' => {
            'address' => '127.0.0.1',
            'port' => 12514
        },
        'fallback_logger' => {
            'file' => STDERR,
            'shift_age' => 0,
            'shift_size' => 1048576
        },
        'backlogLimit' => 512
    }

    @hostname = Socket.gethostname

    def _mergeConfigs(h1, h2)

      h2.each do |key, val|
          if h1.has_key?(key) and val.is_a?(Hash) then
              h1[key] = _mergeConfigs(h1[key], val)
          elsif !h1.has_key?(key)
            h1[key] = val
          end
        end

        return h1;

    end

    @options = _mergeConfigs(opts, @defaults);
    @connected = false
    @socket = nil

    @fallback_logger = Logger.new(@options['fallback_logger']['file'], @options['fallback_logger']['shift_age'], @options['fallback_logger']['shift_size'])

    connect

  end

  def log (message, backtrace = nil, file = nil, line = nil, custom = nil)
    return _log(SirvlogClient::LEVEL_INFO, message, backtrace, file, line, custom);
  end

  def emergency (message, backtrace = nil, file = nil, line = nil, custom = nil)
    return _log(SirvlogClient::LEVEL_EMERGENCY, message, backtrace, file, line, custom);
  end

  def alert (message, backtrace = nil, file = nil, line = nil, custom = nil)
    return _log(SirvlogClient::LEVEL_ALERT, message, backtrace, file, line, custom);
  end

  def critical (message, backtrace = nil, file = nil, line = nil, custom = nil)
    return _log(SirvlogClient::LEVEL_CRITICAL, message, backtrace, file, line, custom);
  end

  def error (message, backtrace = nil, file = nil, line = nil, custom = nil)
    return _log(SirvlogClient::LEVEL_ERROR, message, backtrace, file, line, custom);
  end

  def warning (message, backtrace = nil, file = nil, line = nil, custom = nil)
    return _log(SirvlogClient::LEVEL_WARNING, message, backtrace, file, line, custom);
  end

  def notice (message, backtrace = nil, file = nil, line = nil, custom = nil)
    return _log(SirvlogClient::LEVEL_NOTICE, message, backtrace, file, line, custom);
  end

  def info (message, backtrace = nil, file = nil, line = nil, custom = nil)
    return _log(SirvlogClient::LEVEL_INFO, message, backtrace, file, line, custom);
  end

  def debug (message, backtrace = nil, file = nil, line = nil, custom = nil)
    return _log(SirvlogClient::LEVEL_DEBUG, message, backtrace, file, line, custom);
  end

  def store (logObj)

    if !logObj['hostname']
      logObj['hostname'] = @hostname
    end

    if !logObj['facility']
      logObj['facility'] = @options['facility']
    end

    if !logObj['level']
      logObj['level'] = SirvlogClient::LEVEL_INFO
    end

    json = JSON.generate(logObj)

    if @options['backlogLimit'] then
      if @backlog.count > @options['backlogLimit'] then
        @fallback_logger.warn('backlog buffer is full, dropping message: ' + @backlog.shift)
      end
      @backlog.push(json)
    end

    if !@connected then
      if @options['backlogLimit'] < 1 then
        @fallback_logger.warn('can\'t send message (not connected), dropping: ' + json)
      end
      connect
      return
    end

    if send(json) then
      @backlog.pop
    else
      @fallback_logger.warn("Failed to send message: " + json)
      connect
    end
  end

  private

  def connect
    begin
      @socket = TCPSocket.new @options['server']['address'], @options['server']['port']

      while @backlog.count > 0 do
        json = @backlog[0]
        if send(json) then
          @backlog.shift
        else
          return
        end
      end

      @fallback_logger.debug("Connected to " + @options['server']['address'] + ':' + @options['server']['port'].to_s)
      @connected = true

    rescue Exception => e
      @fallback_logger.fatal(e.message)
    end

  end

  def _log (level, message, backtrace = nil, file = nil, line = nil, custom = nil)

    if level < SirvlogClient::LEVEL_EMERGENCY or level > SirvlogClient::LEVEL_DEBUG then
      level = SirvlogClient::LEVEL_INFO
    end

    if message.length < 1 then
      return
    end

    logObj = {
        'hostname' => @hostname,
        'facility' => @options['facility'],
        'level' => level,
        'message' => message
    }

    if custom.is_a?(Hash) then
        logObj['custom'] = custom
    end

    if line.is_a?(Fixnum) then
        logObj['line'] = line
    elsif line.is_a?(Hash)
        logObj['custom'] = line
    end

    if file.is_a?(String) and file.length > 0 then
        logObj['file'] = file
    elsif file.is_a?(Hash)
        logObj['custom'] = file
    end

    if backtrace.is_a?(String) and backtrace.length > 0 then
        logObj['backtrace'] = backtrace
    elsif backtrace.is_a?(Hash)
        logObj['custom'] = backtrace
    end

    store(logObj)

  end

  def send (message)

    dataLen = message.bytesize

    if dataLen > 65535 then
        @fallback_logger.warn('Message is too long (' + dataLen.to_s + '>65535): ' + message.slice(0, 50) + '...');
        return true
    end

    if dataLen < 512 then
        return _send(SirvlogClient::ENCODING_PLAIN, message)
    else # zlib deflate
        return _send(SirvlogClient::ENCODING_DEFLATE, Zlib::Deflate.deflate(message))
    end

  end

  def _send (encoding, message)
    messsageLen = message.bytesize

    buffer =  [encoding, messsageLen, message].pack('Cna*')

    length = 3 + messsageLen

    while true do

      # check for EOF
      begin
        @socket.read_nonblock(1)
      rescue Errno::EAGAIN
        # nothing
      rescue EOFError
        #@socket.shutdown(:RDWR)
        @socket.shutdown
        @socket.close
        @connected = false
        return false
      rescue Exception => e
        @fallback_logger.fatal(e.message)
        @connected = false
        return false
      end

      begin

        bytesWritten = @socket.send(buffer, 0)

        if bytesWritten < length then
          length -= bytesWritten
          buffer = buffer.slice(bytesWritten, length)
        else
          return true
        end

      rescue Exception => e
        @fallback_logger.fatal(e.message)
        @connected = false
        return false
      end

      return true

    end

  end

end

